File: cpp/cpp_100-999/198.cc
// C++
// leetcode 198
// https://leetcode-cn.com/problems/house-robber/
// 动态规划， int型数组 记录返回值

class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if( n == 0)
            return 0;
        int dp[n + 1];
        dp[0] = 0;
        dp[1] = nums[0];
        for(int i = 2; i <= n; ++i) {
            dp[i] = max(dp[i - 2] + nums[i - 1], dp[i - 1]);
        }
        return dp[n];
    }
};

File: Go/Go_100-999/198.go
// Go
// leetcode 198
// https://leetcode.cn/problems/house-robber/

func rob(nums []int) (res int) {
	pre := make([]int, 3)
	getMaxPreFrom12 := func() int {
		res := 0
		for i := 1; i < 3; i++ {
			res = max(res, pre[i])
		}
		return res
	}

	updatePre := func(x int) {
		pre[2] = pre[1]
		pre[1] = pre[0]
		pre[0] = x
	}

	for _, x := range nums {
		cur := x + getMaxPreFrom12()
		res = max(res, cur)
		updatePre(cur)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}